File: 104_synthetic_control_trading/python/synthetic_control.py
"""
Synthetic Control Method implementation for trading.

Provides the core SCM estimator with:
- Weight optimization using quadratic programming
- Treatment effect estimation
- Placebo test inference
- Visualization utilities
"""

import numpy as np
import logging
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, field

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


@dataclass
class SCMResults:
    """Results from Synthetic Control Method analysis."""
    weights: np.ndarray
    donor_labels: List[str]
    pre_treatment_rmse: float
    treatment_effects: np.ndarray
    cumulative_effect: float
    car: float  # Cumulative abnormal return
    pre_treatment_fit: np.ndarray
    post_treatment_synthetic: np.ndarray


@dataclass
class PlaceboResults:
    """Results from placebo tests."""
    placebo_effects: Dict[str, np.ndarray]
    placebo_rmse: Dict[str, float]
    p_value: float
    rank: int
    n_placebos: int


class SyntheticControlModel:
    """
    Synthetic Control Method model for causal inference in trading.

    This model constructs a synthetic version of a treated unit (e.g., a stock
    affected by an event) using a weighted combination of donor units (similar
    stocks unaffected by the event).

    Example:
        >>> from python.synthetic_control import SyntheticControlModel
        >>> model = SyntheticControlModel()
        >>> model.fit(treated_pre, donors_pre)
        >>> effects = model.estimate_effects(treated_post, donors_post)
        >>> print(f"CAR: {effects['car']:.4f}")
    """

    def __init__(
        self,
        v_penalty: float = 0.0,
        max_weight: float = 1.0,
        min_weight: float = 0.0,
        verbose: bool = False,
    ):
        """
        Initialize Synthetic Control Model.

        Args:
            v_penalty: Regularization penalty for weight optimization
            max_weight: Maximum weight for any single donor
            min_weight: Minimum weight for any donor (usually 0)
            verbose: Whether to print optimization progress
        """
        self.v_penalty = v_penalty
        self.max_weight = max_weight
        self.min_weight = min_weight
        self.verbose = verbose

        self.weights_: Optional[np.ndarray] = None
        self.donor_labels_: Optional[List[str]] = None
        self.pre_treatment_rmse_: Optional[float] = None
        self._treated_pre: Optional[np.ndarray] = None
        self._donors_pre: Optional[np.ndarray] = None

    def fit(
        self,
        treated: np.ndarray,
        donors: np.ndarray,
        donor_labels: Optional[List[str]] = None,
        predictors: Optional[List[str]] = None,
    ) -> "SyntheticControlModel":
        """
        Fit the synthetic control model to pre-treatment data.

        Finds optimal weights W* such that:
            Σⱼ wⱼ * donor_j ≈ treated

        Args:
            treated: Pre-treatment outcomes for treated unit (T_pre,)
            donors: Pre-treatment outcomes for donors (T_pre, J)
            donor_labels: Optional labels for donor units
            predictors: Optional list of predictor names (for documentation)

        Returns:
            Self for method chaining
        """
        treated = np.asarray(treated).flatten()
        donors = np.asarray(donors)

        if donors.ndim == 1:
            donors = donors.reshape(-1, 1)

        T_pre, J = donors.shape

        if len(treated) != T_pre:
            raise ValueError(
                f"Treated length ({len(treated)}) != donors rows ({T_pre})"
            )

        self._treated_pre = treated
        self._donors_pre = donors
        self.donor_labels_ = donor_labels or [f"Donor_{i}" for i in range(J)]

        # Optimize weights using constrained least squares
        self.weights_ = self._optimize_weights(treated, donors)

        # Calculate pre-treatment fit
        synthetic_pre = donors @ self.weights_
        self.pre_treatment_rmse_ = np.sqrt(np.mean((treated - synthetic_pre) ** 2))

        if self.verbose:
            logger.info(f"Pre-treatment RMSE: {self.pre_treatment_rmse_:.6f}")
            for label, weight in zip(self.donor_labels_, self.weights_):
                if weight > 0.01:
                    logger.info(f"  {label}: {weight:.4f}")

        return self

    def _optimize_weights(
        self,
        treated: np.ndarray,
        donors: np.ndarray,
    ) -> np.ndarray:
        """
        Optimize donor weights to minimize pre-treatment prediction error.

        Uses constrained optimization:
        - Weights sum to 1
        - Weights are non-negative (or within [min_weight, max_weight])

        Args:
            treated: Pre-treatment treated outcomes (T,)
            donors: Pre-treatment donor outcomes (T, J)

        Returns:
            Optimal weights (J,)
        """
        T, J = donors.shape

        # Try scipy optimization first, fall back to simple least squares
        try:
            from scipy.optimize import minimize

            def objective(w):
                synthetic = donors @ w
                loss = np.sum((treated - synthetic) ** 2)
                # Add L2 regularization
                loss += self.v_penalty * np.sum(w ** 2)
                return loss

            # Constraints: weights sum to 1
            constraints = {"type": "eq", "fun": lambda w: np.sum(w) - 1.0}

            # Bounds: weights in [min_weight, max_weight]
            bounds = [(self.min_weight, self.max_weight) for _ in range(J)]

            # Initial guess: equal weights
            w0 = np.ones(J) / J

            result = minimize(
                objective,
                w0,
                method="SLSQP",
                bounds=bounds,
                constraints=constraints,
                options={"maxiter": 1000, "ftol": 1e-10},
            )

            if result.success:
                weights = result.x
                # Ensure constraints are satisfied
                weights = np.maximum(weights, self.min_weight)
                weights = np.minimum(weights, self.max_weight)
                weights = weights / np.sum(weights)
                return weights
            else:
                logger.warning(f"Optimization failed: {result.message}")

        except ImportError:
            logger.warning("scipy not available, using simple least squares")

        # Fallback: Simple least squares with projection
        weights = self._simple_ols_weights(treated, donors)
        return weights

    def _simple_ols_weights(
        self,
        treated: np.ndarray,
        donors: np.ndarray,
    ) -> np.ndarray:
        """
        Simple OLS-based weight estimation with projection to constraints.

        Args:
            treated: Pre-treatment treated outcomes (T,)
            donors: Pre-treatment donor outcomes (T, J)

        Returns:
            Weights projected to simplex (J,)
        """
        # OLS solution
        try:
            weights = np.linalg.lstsq(donors, treated, rcond=None)[0]
        except np.linalg.LinAlgError:
            weights = np.ones(donors.shape[1]) / donors.shape[1]

        # Project to simplex (non-negative, sum to 1)
        weights = self._project_to_simplex(weights)
        return weights

    def _project_to_simplex(self, v: np.ndarray) -> np.ndarray:
        """
        Project vector onto probability simplex.

        Args:
            v: Input vector

        Returns:
            Projected vector (non-negative, sums to 1)
        """
        n = len(v)
        u = np.sort(v)[::-1]
        cssv = np.cumsum(u) - 1
        ind = np.arange(1, n + 1)
        cond = u - cssv / ind > 0
        rho = ind[cond][-1]
        theta = cssv[cond][-1] / rho
        w = np.maximum(v - theta, 0)
        return w

    def estimate_effects(
        self,
        treated_post: np.ndarray,
        donors_post: np.ndarray,
    ) -> Dict:
        """
        Estimate treatment effects using fitted weights.

        Args:
            treated_post: Post-treatment outcomes for treated unit (T_post,)
            donors_post: Post-treatment outcomes for donors (T_post, J)

        Returns:
            Dictionary with:
            - effects: Point-by-point treatment effects
            - cumulative: Cumulative treatment effect
            - car: Cumulative abnormal return
            - synthetic_post: Synthetic control post-treatment values
        """
        if self.weights_ is None:
            raise RuntimeError("Model must be fitted before estimating effects")

        treated_post = np.asarray(treated_post).flatten()
        donors_post = np.asarray(donors_post)

        if donors_post.ndim == 1:
            donors_post = donors_post.reshape(-1, 1)

        # Construct synthetic control for post-treatment period
        synthetic_post = donors_post @ self.weights_

        # Treatment effects (gap)
        effects = treated_post - synthetic_post

        # Cumulative effect
        cumulative_effect = np.sum(effects)

        # CAR (for returns data, this is the cumulative abnormal return)
        car = cumulative_effect

        return {
            "effects": effects,
            "cumulative": cumulative_effect,
            "car": car,
            "synthetic_post": synthetic_post,
            "treated_post": treated_post,
        }

    def get_results(
        self,
        treated_post: np.ndarray,
        donors_post: np.ndarray,
    ) -> SCMResults:
        """
        Get comprehensive SCM results.

        Args:
            treated_post: Post-treatment outcomes for treated unit
            donors_post: Post-treatment outcomes for donors

        Returns:
            SCMResults dataclass with all analysis outputs
        """
        effects = self.estimate_effects(treated_post, donors_post)

        pre_fit = self._donors_pre @ self.weights_

        return SCMResults(
            weights=self.weights_,
            donor_labels=self.donor_labels_,
            pre_treatment_rmse=self.pre_treatment_rmse_,
            treatment_effects=effects["effects"],
            cumulative_effect=effects["cumulative"],
            car=effects["car"],
            pre_treatment_fit=pre_fit,
            post_treatment_synthetic=effects["synthetic_post"],
        )

    def weights(self) -> Dict[str, float]:
        """Get donor weights as a dictionary."""
        if self.weights_ is None:
            raise RuntimeError("Model must be fitted first")
        return dict(zip(self.donor_labels_, self.weights_))

    def run_placebo_tests(
        self,
        treated_post: np.ndarray,
        donors_post: np.ndarray,
    ) -> PlaceboResults:
        """
        Run in-space placebo tests for inference.

        For each donor unit, pretend it was the treated unit and
        estimate a placebo treatment effect. The true treatment
        effect is significant if it is extreme relative to placebos.

        Args:
            treated_post: Post-treatment outcomes for treated unit
            donors_post: Post-treatment outcomes for donors

        Returns:
            PlaceboResults with p-value and rank
        """
        if self.weights_ is None:
            raise RuntimeError("Model must be fitted first")

        treated_post = np.asarray(treated_post).flatten()
        donors_post = np.asarray(donors_post)

        if donors_post.ndim == 1:
            donors_post = donors_post.reshape(-1, 1)

        # Get actual treatment effect
        actual_effects = self.estimate_effects(treated_post, donors_post)
        actual_car = np.abs(actual_effects["car"])

        # Run placebos
        placebo_effects = {}
        placebo_rmse = {}
        placebo_cars = []

        J = self._donors_pre.shape[1]

        for j in range(J):
            # Donor j as treated, others as donors
            placebo_treated_pre = self._donors_pre[:, j]
            placebo_donors_pre = np.delete(self._donors_pre, j, axis=1)

            placebo_treated_post = donors_post[:, j]
            placebo_donors_post = np.delete(donors_post, j, axis=1)

            # Fit placebo model
            placebo_model = SyntheticControlModel(
                v_penalty=self.v_penalty,
                max_weight=self.max_weight,
                min_weight=self.min_weight,
            )
            placebo_model.fit(placebo_treated_pre, placebo_donors_pre)

            # Only include if pre-treatment fit is reasonable
            if placebo_model.pre_treatment_rmse_ < 2 * self.pre_treatment_rmse_:
                effects = placebo_model.estimate_effects(
                    placebo_treated_post, placebo_donors_post
                )
                label = self.donor_labels_[j]
                placebo_effects[label] = effects["effects"]
                placebo_rmse[label] = placebo_model.pre_treatment_rmse_
                placebo_cars.append(np.abs(effects["car"]))

        # Calculate p-value (rank-based)
        all_cars = placebo_cars + [actual_car]
        rank = sum(1 for x in all_cars if x >= actual_car)
        p_value = rank / len(all_cars)

        return PlaceboResults(
            placebo_effects=placebo_effects,
            placebo_rmse=placebo_rmse,
            p_value=p_value,
            rank=rank,
            n_placebos=len(placebo_cars),
        )


def run_scm_analysis(
    treated_pre: np.ndarray,
    donors_pre: np.ndarray,
    treated_post: np.ndarray,
    donors_post: np.ndarray,
    donor_labels: Optional[List[str]] = None,
    run_placebos: bool = True,
) -> Tuple[SCMResults, Optional[PlaceboResults]]:
    """
    Run complete SCM analysis.

    Convenience function that fits the model, estimates effects,
    and optionally runs placebo tests.

    Args:
        treated_pre: Pre-treatment treated outcomes
        donors_pre: Pre-treatment donor outcomes
        treated_post: Post-treatment treated outcomes
        donors_post: Post-treatment donor outcomes
        donor_labels: Optional donor labels
        run_placebos: Whether to run placebo tests

    Returns:
        Tuple of (SCMResults, PlaceboResults or None)
    """
    model = SyntheticControlModel(verbose=True)
    model.fit(treated_pre, donors_pre, donor_labels)

    results = model.get_results(treated_post, donors_post)

    placebo_results = None
    if run_placebos:
        placebo_results = model.run_placebo_tests(treated_post, donors_post)

    return results, placebo_results

File: 104_synthetic_control_trading/src/model/synthetic_control.rs
//! Synthetic Control Method implementation.
//!
//! Provides the core SCM estimator with:
//! - Weight optimization
//! - Treatment effect estimation
//! - Placebo test inference

use anyhow::{anyhow, Result};
use std::collections::HashMap;

/// Results from Synthetic Control Method analysis.
#[derive(Debug, Clone)]
pub struct SCMResults {
    /// Optimal donor weights
    pub weights: Vec<f64>,
    /// Donor labels
    pub donor_labels: Vec<String>,
    /// Pre-treatment RMSE
    pub pre_treatment_rmse: f64,
    /// Treatment effects over time
    pub treatment_effects: Vec<f64>,
    /// Cumulative treatment effect
    pub cumulative_effect: f64,
    /// Cumulative abnormal return
    pub car: f64,
    /// Pre-treatment fitted values
    pub pre_treatment_fit: Vec<f64>,
    /// Post-treatment synthetic values
    pub post_treatment_synthetic: Vec<f64>,
}

/// Results from placebo tests.
#[derive(Debug, Clone)]
pub struct PlaceboResults {
    /// Placebo effects by donor
    pub placebo_effects: HashMap<String, Vec<f64>>,
    /// P-value from permutation test
    pub p_value: f64,
    /// Rank of actual effect among placebos
    pub rank: usize,
    /// Number of placebos run
    pub n_placebos: usize,
}

/// Synthetic Control Method model for causal inference in trading.
///
/// This model constructs a synthetic version of a treated unit using
/// a weighted combination of donor units.
///
/// # Example
///
/// ```rust,no_run
/// use synthetic_control_trading::SyntheticControlModel;
///
/// let mut model = SyntheticControlModel::new();
/// let treated_pre = vec![0.01, 0.02, -0.01, 0.015];
/// let donors_pre = vec![
///     vec![0.008, 0.018, -0.012, 0.014],
///     vec![0.012, 0.022, -0.008, 0.016],
/// ];
/// model.fit(&treated_pre, &donors_pre).unwrap();
/// ```
#[derive(Debug, Clone)]
pub struct SyntheticControlModel {
    /// Regularization penalty
    v_penalty: f64,
    /// Maximum weight for any donor
    max_weight: f64,
    /// Minimum weight for any donor
    min_weight: f64,
    /// Fitted weights
    weights: Option<Vec<f64>>,
    /// Donor labels
    donor_labels: Option<Vec<String>>,
    /// Pre-treatment RMSE
    pre_treatment_rmse: Option<f64>,
    /// Stored pre-treatment data
    treated_pre: Option<Vec<f64>>,
    donors_pre: Option<Vec<Vec<f64>>>,
}

impl Default for SyntheticControlModel {
    fn default() -> Self {
        Self::new()
    }
}

impl SyntheticControlModel {
    /// Create a new Synthetic Control Model.
    pub fn new() -> Self {
        Self {
            v_penalty: 0.0,
            max_weight: 1.0,
            min_weight: 0.0,
            weights: None,
            donor_labels: None,
            pre_treatment_rmse: None,
            treated_pre: None,
            donors_pre: None,
        }
    }

    /// Set regularization penalty.
    pub fn with_penalty(mut self, penalty: f64) -> Self {
        self.v_penalty = penalty;
        self
    }

    /// Set weight bounds.
    pub fn with_bounds(mut self, min_weight: f64, max_weight: f64) -> Self {
        self.min_weight = min_weight;
        self.max_weight = max_weight;
        self
    }

    /// Fit the synthetic control model to pre-treatment data.
    ///
    /// Finds optimal weights W* such that: Σⱼ wⱼ * donor_j ≈ treated
    ///
    /// # Arguments
    ///
    /// * `treated` - Pre-treatment outcomes for treated unit
    /// * `donors` - Pre-treatment outcomes for each donor (outer vec = donors, inner vec = time)
    ///
    /// # Returns
    ///
    /// Result indicating success or failure
    pub fn fit(&mut self, treated: &[f64], donors: &[Vec<f64>]) -> Result<()> {
        if donors.is_empty() {
            return Err(anyhow!("At least one donor is required"));
        }

        let n_donors = donors.len();
        let n_periods = treated.len();

        // Validate donor lengths
        for (i, donor) in donors.iter().enumerate() {
            if donor.len() != n_periods {
                return Err(anyhow!(
                    "Donor {} has {} periods, expected {}",
                    i,
                    donor.len(),
                    n_periods
                ));
            }
        }

        // Store data
        self.treated_pre = Some(treated.to_vec());
        self.donors_pre = Some(donors.to_vec());
        self.donor_labels = Some((0..n_donors).map(|i| format!("Donor_{}", i)).collect());

        // Optimize weights
        self.weights = Some(self.optimize_weights(treated, donors)?);

        // Calculate pre-treatment fit
        let synthetic_pre = self.compute_synthetic(treated.len(), donors)?;
        let mse: f64 = treated
            .iter()
            .zip(synthetic_pre.iter())
            .map(|(t, s)| (t - s).powi(2))
            .sum::<f64>()
            / n_periods as f64;
        self.pre_treatment_rmse = Some(mse.sqrt());

        Ok(())
    }

    /// Fit with custom donor labels.
    pub fn fit_with_labels(
        &mut self,
        treated: &[f64],
        donors: &[Vec<f64>],
        labels: Vec<String>,
    ) -> Result<()> {
        self.fit(treated, donors)?;
        self.donor_labels = Some(labels);
        Ok(())
    }

    /// Optimize donor weights using constrained least squares.
    fn optimize_weights(&self, treated: &[f64], donors: &[Vec<f64>]) -> Result<Vec<f64>> {
        let n_donors = donors.len();
        let n_periods = treated.len();

        // Simple iterative optimization (gradient descent with projection)
        let mut weights = vec![1.0 / n_donors as f64; n_donors];
        let learning_rate = 0.1;
        let max_iterations = 1000;
        let tolerance = 1e-8;

        for _ in 0..max_iterations {
            // Compute synthetic
            let synthetic: Vec<f64> = (0..n_periods)
                .map(|t| {
                    donors
                        .iter()
                        .zip(weights.iter())
                        .map(|(d, w)| d[t] * w)
                        .sum()
                })
                .collect();

            // Compute gradients
            let mut gradients = vec![0.0; n_donors];
            for (j, donor) in donors.iter().enumerate() {
                for t in 0..n_periods {
                    gradients[j] += 2.0 * (synthetic[t] - treated[t]) * donor[t];
                }
                gradients[j] /= n_periods as f64;
                gradients[j] += 2.0 * self.v_penalty * weights[j];
            }

            // Update weights
            let mut new_weights = weights.clone();
            for (j, grad) in gradients.iter().enumerate() {
                new_weights[j] -= learning_rate * grad;
            }

            // Project to simplex
            new_weights = self.project_to_simplex(&new_weights);

            // Check convergence
            let diff: f64 = new_weights
                .iter()
                .zip(weights.iter())
                .map(|(n, o)| (n - o).abs())
                .sum();

            weights = new_weights;

            if diff < tolerance {
                break;
            }
        }

        Ok(weights)
    }

    /// Project vector onto probability simplex.
    fn project_to_simplex(&self, v: &[f64]) -> Vec<f64> {
        let mut u: Vec<f64> = v.to_vec();
        u.sort_by(|a, b| b.partial_cmp(a).unwrap());

        let mut cssv = 0.0;
        let mut rho = 0;

        for (i, &ui) in u.iter().enumerate() {
            cssv += ui;
            if ui - (cssv - 1.0) / (i + 1) as f64 > 0.0 {
                rho = i + 1;
            }
        }

        let theta = (u[..rho].iter().sum::<f64>() - 1.0) / rho as f64;

        v.iter().map(|&vi| (vi - theta).max(0.0)).collect()
    }

    /// Compute synthetic control values.
    fn compute_synthetic(&self, n_periods: usize, donors: &[Vec<f64>]) -> Result<Vec<f64>> {
        let weights = self
            .weights
            .as_ref()
            .ok_or_else(|| anyhow!("Model not fitted"))?;

        let synthetic: Vec<f64> = (0..n_periods)
            .map(|t| {
                donors
                    .iter()
                    .zip(weights.iter())
                    .map(|(d, w)| {
                        if t < d.len() {
                            d[t] * w
                        } else {
                            0.0
                        }
                    })
                    .sum()
            })
            .collect();

        Ok(synthetic)
    }

    /// Estimate treatment effects using fitted weights.
    ///
    /// # Arguments
    ///
    /// * `treated_post` - Post-treatment outcomes for treated unit
    /// * `donors_post` - Post-treatment outcomes for donors
    ///
    /// # Returns
    ///
    /// Dictionary-like structure with effects, CAR, etc.
    pub fn estimate_effects(
        &self,
        treated_post: &[f64],
        donors_post: &[Vec<f64>],
    ) -> Result<SCMEffects> {
        let weights = self
            .weights
            .as_ref()
            .ok_or_else(|| anyhow!("Model not fitted"))?;

        let n_post = treated_post.len();

        // Compute synthetic post-treatment
        let synthetic_post: Vec<f64> = (0..n_post)
            .map(|t| {
                donors_post
                    .iter()
                    .zip(weights.iter())
                    .map(|(d, w)| {
                        if t < d.len() {
                            d[t] * w
                        } else {
                            0.0
                        }
                    })
                    .sum()
            })
            .collect();

        // Treatment effects (gap)
        let effects: Vec<f64> = treated_post
            .iter()
            .zip(synthetic_post.iter())
            .map(|(t, s)| t - s)
            .collect();

        let cumulative: f64 = effects.iter().sum();

        Ok(SCMEffects {
            effects,
            cumulative,
            car: cumulative,
            synthetic_post,
            treated_post: treated_post.to_vec(),
        })
    }

    /// Get the fitted weights.
    pub fn weights(&self) -> Option<&Vec<f64>> {
        self.weights.as_ref()
    }

    /// Get weights as a hashmap with labels.
    pub fn weights_map(&self) -> Option<HashMap<String, f64>> {
        match (&self.weights, &self.donor_labels) {
            (Some(w), Some(l)) => Some(l.iter().cloned().zip(w.iter().cloned()).collect()),
            _ => None,
        }
    }

    /// Get pre-treatment RMSE.
    pub fn pre_treatment_rmse(&self) -> Option<f64> {
        self.pre_treatment_rmse
    }

    /// Get comprehensive results.
    pub fn get_results(
        &self,
        treated_post: &[f64],
        donors_post: &[Vec<f64>],
    ) -> Result<SCMResults> {
        let effects = self.estimate_effects(treated_post, donors_post)?;

        let pre_fit = self.compute_synthetic(
            self.treated_pre.as_ref().map_or(0, |v| v.len()),
            self.donors_pre.as_ref().unwrap_or(&vec![]),
        )?;

        Ok(SCMResults {
            weights: self.weights.clone().unwrap_or_default(),
            donor_labels: self.donor_labels.clone().unwrap_or_default(),
            pre_treatment_rmse: self.pre_treatment_rmse.unwrap_or(0.0),
            treatment_effects: effects.effects,
            cumulative_effect: effects.cumulative,
            car: effects.car,
            pre_treatment_fit: pre_fit,
            post_treatment_synthetic: effects.synthetic_post,
        })
    }

    /// Run in-space placebo tests for inference.
    pub fn run_placebo_tests(
        &self,
        treated_post: &[f64],
        donors_post: &[Vec<f64>],
    ) -> Result<PlaceboResults> {
        let donors_pre = self
            .donors_pre
            .as_ref()
            .ok_or_else(|| anyhow!("Model not fitted"))?;

        // Get actual treatment effect
        let actual_effects = self.estimate_effects(treated_post, donors_post)?;
        let actual_car = actual_effects.car.abs();

        let n_donors = donors_pre.len();
        let mut placebo_effects: HashMap<String, Vec<f64>> = HashMap::new();
        let mut placebo_cars: Vec<f64> = Vec::new();

        for j in 0..n_donors {
            // Donor j as treated, others as donors
            let placebo_treated_pre = &donors_pre[j];
            let placebo_donors_pre: Vec<Vec<f64>> = donors_pre
                .iter()
                .enumerate()
                .filter(|(i, _)| *i != j)
                .map(|(_, d)| d.clone())
                .collect();

            let placebo_treated_post = &donors_post[j];
            let placebo_donors_post: Vec<Vec<f64>> = donors_post
                .iter()
                .enumerate()
                .filter(|(i, _)| *i != j)
                .map(|(_, d)| d.clone())
                .collect();

            // Fit placebo model
            let mut placebo_model = SyntheticControlModel::new()
                .with_penalty(self.v_penalty)
                .with_bounds(self.min_weight, self.max_weight);

            if placebo_model
                .fit(placebo_treated_pre, &placebo_donors_pre)
                .is_ok()
            {
                // Only include if pre-treatment fit is reasonable
                if let Some(rmse) = placebo_model.pre_treatment_rmse() {
                    if rmse < 2.0 * self.pre_treatment_rmse.unwrap_or(f64::MAX) {
                        if let Ok(effects) =
                            placebo_model.estimate_effects(placebo_treated_post, &placebo_donors_post)
                        {
                            let label = format!("Donor_{}", j);
                            placebo_effects.insert(label, effects.effects);
                            placebo_cars.push(effects.car.abs());
                        }
                    }
                }
            }
        }

        // Calculate p-value (rank-based)
        let rank = placebo_cars.iter().filter(|&&c| c >= actual_car).count() + 1;
        let p_value = rank as f64 / (placebo_cars.len() + 1) as f64;

        Ok(PlaceboResults {
            placebo_effects,
            p_value,
            rank,
            n_placebos: placebo_cars.len(),
        })
    }
}

/// Effects from SCM analysis.
#[derive(Debug, Clone)]
pub struct SCMEffects {
    /// Point-by-point treatment effects
    pub effects: Vec<f64>,
    /// Cumulative treatment effect
    pub cumulative: f64,
    /// Cumulative abnormal return
    pub car: f64,
    /// Synthetic post-treatment values
    pub synthetic_post: Vec<f64>,
    /// Actual post-treatment values
    pub treated_post: Vec<f64>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fit_basic() {
        let mut model = SyntheticControlModel::new();

        let treated = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let donors = vec![
            vec![1.1, 2.1, 3.1, 4.1, 5.1],
            vec![0.9, 1.9, 2.9, 3.9, 4.9],
        ];

        model.fit(&treated, &donors).unwrap();

        assert!(model.weights().is_some());
        assert!(model.pre_treatment_rmse().is_some());
    }

    #[test]
    fn test_estimate_effects() {
        let mut model = SyntheticControlModel::new();

        let treated_pre = vec![1.0, 2.0, 3.0, 4.0];
        let donors_pre = vec![
            vec![1.0, 2.0, 3.0, 4.0],
            vec![1.0, 2.0, 3.0, 4.0],
        ];

        model.fit(&treated_pre, &donors_pre).unwrap();

        let treated_post = vec![5.5, 6.5]; // Higher than expected
        let donors_post = vec![vec![5.0, 6.0], vec![5.0, 6.0]];

        let effects = model.estimate_effects(&treated_post, &donors_post).unwrap();

        assert!(effects.car > 0.0); // Positive treatment effect
    }

    #[test]
    fn test_project_to_simplex() {
        let model = SyntheticControlModel::new();

        let v = vec![0.5, 0.3, 0.3];
        let projected = model.project_to_simplex(&v);

        let sum: f64 = projected.iter().sum();
        assert!((sum - 1.0).abs() < 1e-6);
        assert!(projected.iter().all(|&x| x >= 0.0));
    }
}
